android/runtime/common/src/java/org/appcelerator/kroll/util/KrollAssetHelper.java

```diff
@@ -114,21 +184,63 @@ public static String readFile(String path)
 
 	public static boolean assetExists(String path)
 	{
-		if (assetCrypt != null) {
-			String asset = assetCrypt.readAsset(path.replace("Resources/", ""));
-			if (asset != null) {
+		// Validate argument.
+		if (path == null) {
+			return false;
+		}
+
+		// First, check if path references an encrypted Titanium asset.
+		Collection<String> collection = getEncryptedAssetPaths();
+		if (collection != null) {
+			if (collection.contains(normalizeAssetCryptPath(path))) {
 				return true;
 			}
 		}
-		if (assetManager != null) {
-			try {
-				return assetManager != null && assetManager.open(path) != null;
-			} catch (IOException e) {
-			}
+
+		// Next, check if path exists under APK "assets".
+		if (apkAssetFilePathSet.contains(path)) {
+			return true;
+		}
+
+		// Asset path not found.
+		return false;
+	}
+
+	public static boolean directoryExists(String path)
+	{
+		path = normalizeDirectoryPath(path);
+		if (encrpytedDirectoryListingMap.containsKey(normalizeAssetCryptPath(path))) {
+			return true;
+		}
+		if (apkDirectoryListingMap.containsKey(path)) {
+			return true;
 		}
 		return false;
 	}
 
+	public static List<String> getDirectoryListing(String path)
+	{
+		ArrayList<String> resultList = new ArrayList<>();
+		path = normalizeDirectoryPath(path);
+		{
+			Collection<String> collection = encrpytedDirectoryListingMap.get(normalizeAssetCryptPath(path));
+			if (collection != null) {
+				resultList.addAll(collection);
+			}
+		}
+		{
+			Collection<String> collection = apkDirectoryListingMap.get(path);
+			if (collection != null) {
+				for (String entry : collection) {
+					if (!resultList.contains(entry)) {
```

Comment:
Maybe we could swap the order of this method, so we are doing the `!resultList.contains(entry)` on the most likely smaller collection from `encrpytedDirectoryListingMap` ?

Just a thought I had

Reply:
Good idea. Especially since the `ArrayList.contains()` is an `O(n)` lookup.